### PR TITLE
Scope added to autolink filter

### PR DESCRIPTION
**Pull request by scottleedavis**
Adds Team and Channel Scope to plugin config, and restricts autolinks under given Scope if present.

Contributing for https://github.com/mattermost/mattermost-plugin-autolink/issues/10

### CODE REVIEW

**Comment by hanzei** [comment]
What do you think of consolidating TeamScope and ChannelScope inside one scope and letting it have it be like `$TEAM/$CHANNEL`. Otherwise you can run into problems like if a channel is present in two teams.

**Comment by scottleedavis** [comment]
That could work, though what if it was desired to scope by channel name and not team?

**Comment by hanzei** [comment]
I feel this is a rare use case. But then you can just add a scope for every team.

File: README.md [hunk 1]
```diff
@@ -28,11 +28,13 @@ Below is an example of regexp patterns used for autolinking, modified in the `co
             "links": [
                 {
                     "Pattern": "(LHS)",
-                    "Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)"
+                    "Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)",
+                    "Scope": ["test/off-topic"]
```
Comment (hanzei) [inline README.md]: Just to make the example clear:
```suggestion
                    "Scope": ["team1/off-topic"]
```

File: README.md [hunk 2]
```diff
@@ -28,11 +28,13 @@ Below is an example of regexp patterns used for autolinking, modified in the `co
...
                 },
                 {
                     "Pattern": "(RHS)",
-                    "Template": "[RHS](https://docs.mattermost.com/process/training.html#rhs)"
+                    "Template": "[RHS](https://docs.mattermost.com/process/training.html#rhs)",
+                    "Scope": ["another-test/town-square"]
```
Comment (hanzei) [inline README.md]: Let's add an example for a team scope
```suggestion
                    "Scope": ["team2"]
```

File: server/plugin.go [hunk 3]
```diff
@@ -82,10 +82,17 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
 
 		if origText != "" {
 			newText := origText
+
+			channel, _ := p.API.GetChannel(post.ChannelId)
```
Comment (hanzei) [inline server/plugin.go]: 3/5 that you should not ignore the error here.

File: server/plugin.go [hunk 4]
```diff
@@ -82,10 +82,17 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
...
+			team, _ := p.API.GetTeam(channel.TeamId)
```
Comment (hanzei) [inline server/plugin.go]: Same here.

File: server/plugin.go [hunk 5]
```diff
@@ -98,3 +105,16 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
 
 	return post, ""
 }
+
+func contains(channel string, team string, list []string) bool {
```
Comment (hanzei) [inline server/plugin.go]: Mind moving this into a `util.go` file?

File: server/plugin.go [hunk 6]
```diff
@@ -82,10 +82,17 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
...
 			newText := origText
+
+			channel, _ := p.API.GetChannel(post.ChannelId)
+			team, _ := p.API.GetTeam(channel.TeamId)
+
 			for _, l := range links {
-				newText = l.Replace(newText)
+				if len(l.link.Scope) == 0 {
+					newText = l.Replace(newText)
+				} else if contains(channel.Name, team.Name, l.link.Scope) {
```
Comment (hanzei) [inline server/plugin.go]: It's kind of hard to read. That do you think about using 
```suggestion
				if len(l.link.Scope) != 0 && !contains(channel.Name, team.Name, l.link.Scope) {
					continue
				}
				newText = l.Replace(newText)
```

instead?

File: server/plugin.go [hunk 7]
```diff
@@ -98,3 +105,16 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
 
 	return post, ""
 }
+
+func contains(channel string, team string, list []string) bool {
+	for _, channelTeam := range list {
+		channelTeamSplit := strings.Split(channelTeam, "/")
+		if len(channelTeamSplit) == 2 {
```
Comment (hanzei) [inline server/plugin.go]: Maybe log an error if `len(channelTeamSplit) != 2`

File: server/plugin.go [hunk 8]
```diff
@@ -98,3 +105,16 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
...
+			if strings.EqualFold(channelTeamSplit[0], team) && strings.EqualFold(channelTeamSplit[1], channel) {
+				return true
+			}
+		}
+
```
Comment (hanzei) [inline server/plugin.go]: Nit: Remove this line

File: README.md (hunk 2 above)
Comment (hanzei) [reply README.md]: Mind adding an example for multiple scope per rule?

File: server/plugin.go (hunk 6 above)
Comment (scottleedavis) [reply server/plugin.go]: If there is no Scope provided in configuration, == 0 is a valid check.

Let me know if I missed something.

File: server/plugin.go (hunk 7 above)
Comment (scottleedavis) [reply server/plugin.go]: logging for != 1 & != 2.  (1 being the team only example you provided earlier)

File: server/plugin.go (hunk 6 above)
Comment (hanzei) [reply server/plugin.go]: Maybe we misunderstand each other.

I'm proposing to check first, if a scope is set and then if the scope is meet. The default case should be to apply the replacer.

File: server/plugin.go [hunk 9]
```diff
@@ -83,8 +82,14 @@ func (p *Plugin) MessageWillBePosted(c *plugin.Context, post *model.Post) (*mode
 		if origText != "" {
 			newText := origText
 
-			channel, _ := p.API.GetChannel(post.ChannelId)
-			team, _ := p.API.GetTeam(channel.TeamId)
+			channel, cErr := p.API.GetChannel(post.ChannelId)
+			if cErr != nil {
+				return false
```
Comment (hanzei) [inline server/plugin.go]: Please log an error.

File: server/plugin.go (hunk 6 above)
Comment (scottleedavis) [reply server/plugin.go]: Will you block on this?

Comment (hanzei) [reply server/plugin.go]: If you are at least 4/5, that you code is well the current way, no. I'm just 2/5.

Comment (scottleedavis) [reply server/plugin.go]: I feel that handling 

No Scope in configuration
or
Scope contains team & channel
or 
Scope contains team

I am 4/5 I would change it if you found a more elegant solution to the above logic.

Comment (hanzei) [reply server/plugin.go]: Could you please point out, what is wrong on the solution pointed out above?

Comment (hanzei) [reply server/plugin.go]: Sorry, but now I want ambitious to figure out "the right solution".

File: server/plugin_test.go [hunk 10]
```diff
@@ -49,16 +60,35 @@ func TestSpecialCases(t *testing.T) {
 	}, &Link{
 		Pattern:  "(foo!bar)",
 		Template: "fb",
+	}, &Link{
+		Pattern: "(example)",
+		Template: "test",
```
Comment (scottleedavis) [inline server/plugin_test.go]: should change to team

File: server/plugin_test.go [hunk 11]
```diff
@@ -49,16 +60,35 @@ func TestSpecialCases(t *testing.T) {
...
+		Scope: []string{"test/off-topic"},
+	}, &Link{
+		Pattern: "(example)",
+		Template: "test",
```
Comment (scottleedavis) [inline server/plugin_test.go]: also team

File: README.md [hunk 12]
```diff
@@ -28,11 +28,13 @@ Below is an example of regexp patterns used for autolinking, modified in the `co
...
                     "Pattern": "(LHS)",
-                    "Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)"
+                    "Template": "[LHS](https://docs.mattermost.com/process/training.html#lhs)",
+                    "Scope": ["team1/off-topic"]
                 },
                 {
                     "Pattern": "(RHS)",
-                    "Template": "[RHS](https://docs.mattermost.com/process/training.html#rhs)"
+                    "Template": "[RHS](https://docs.mattermost.com/process/training.html#rhs)",
+                    "Scope": ["team2", "team3", "team4/town-square"]
```
Comment (scottleedavis) [inline README.md]: I think I put this here while thinking about what I wanted, without adding a note to implement it.

File: server/util.go [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+	"strings"
+
+	"github.com/mattermost/mattermost-server/mlog"
+)
+
+func contains(team string, channel string, list []string) bool {
+	for _, channelTeam := range list {
+		channelTeamSplit := strings.Split(channelTeam, "/")
+		if len(channelTeamSplit) == 2 {
+			if strings.EqualFold(channelTeamSplit[0], team) && strings.EqualFold(channelTeamSplit[1], channel) {
```
Comment (scottleedavis) [inline server/util.go]: need to check for ',' string arrays.

File: README.md (hunk 12 above)
Comment (scottleedavis) [reply README.md]: actually, i don't like it.  removing.

File: server/manifest.go [hunk 14]
```diff
@@ -5,5 +5,5 @@ var manifest = struct {
 	Version string
 }{
 	Id:      "mattermost-autolink",
-	Version: "0.4.0",
+	Version: "0.5.0",
```
Comment (hanzei) [inline server/manifest.go]: Please revert this. I like to have things clean and use a seperate PR for this.

**Pull request merged**